oasislmf/pytools/gulmc/manager.py

```diff
@@ -0,0 +1,886 @@
+            corr_data_by_item_id[0] = (0, 0.)
+            corr_data_by_item_id[1:]['peril_correlation_group'] = np.array(data['peril_correlation_group'])
+            corr_data_by_item_id[1:]['correlation_value'] = np.array(data['correlation_value'])
+
+            logger.info(
+                f"Correlation values for {Nperil_correlation_groups} peril correlation groups have been imported."
+            )
+
+            unique_peril_correlation_groups = np.unique(corr_data_by_item_id[1:]['peril_correlation_group'])
+
+            # pre-compute lookup tables for the Gaussian cdf and inverse cdf
+            # Notes:
+            #  - the size `arr_N` and `arr_N_cdf` can be increased to achieve better resolution in the Gaussian cdf and inv cdf.
+            #  - the function `get_corr_rval` to compute the correlated numbers is not affected by arr_N and arr_N_cdf
+            arr_min, arr_max, arr_N = 1e-16, 1 - 1e-16, 1000000
+            arr_min_cdf, arr_max_cdf, arr_N_cdf = -20., 20., 1000000
+            norm_inv_cdf = compute_norm_inv_cdf_lookup(arr_min, arr_max, arr_N)
+            norm_cdf = compute_norm_cdf_lookup(arr_min_cdf, arr_max_cdf, arr_N_cdf)
+
+            # buffer to be re-used to store all the correlated random values
+            z_unif = np.zeros(sample_size, dtype='float64')
+
+        else:
+            # create dummy data structures with proper dtypes to allow correct numba compilation
+            corr_data_by_item_id = np.ndarray(1, dtype=Correlation)
+            arr_min, arr_max, arr_N = 0, 0, 0
+            arr_min_cdf, arr_max_cdf, arr_N_cdf = 0, 0, 0
+            norm_inv_cdf, norm_cdf = np.zeros(1, dtype='float64'), np.zeros(1, dtype='float64')
+            z_unif = np.zeros(1, dtype='float64')
+
+        if effective_damageability is True:
+            logger.info("effective_damageability is False: gulmc will perform the full Monte Carlo sampling: "
+                        "sample the hazard intensity first, then sample the damage from the corresponding vulnerability function.")
+        else:
+            logger.info("effective_damageability is True: gulmc will draw the damage samples from the effective damageability distribution.")
+
+        # create buffers to be reused when computing losses
+        losses = np.zeros((sample_size + NUM_IDX + 1, np.max(coverages[1:]['max_items'])), dtype=oasis_float)
+        vuln_prob_to = np.zeros(Ndamage_bins_max, dtype=oasis_float)
+
+        # maximum bytes to be written in the output stream for 1 item
+        max_bytes_per_item = (sample_size + NUM_IDX + 1) * gulSampleslevelRec_size + 2 * gulSampleslevelHeader_size
+
+        while True:
+            if not streams_in.readinto(event_id_mv):
+                break
+
+            # get the next event_id from the input stream
+            event_id = event_ids[0]
+
+            event_footprint = (FootprintLayerClient if data_server else footprint_obj).get_event(event_id)
```

Comment:
not massively important but I would set the object outside the loop to avoid the if each time

Reply:
good catch, it definitely goes outside the loop

Reply:
done